src/tiered/tiered_handle.c

```diff
@@ -740,6 +740,19 @@ __tiered_open(WT_SESSION_IMPL *session, const char *cfg[])
         if (session->import_list != NULL)
             WT_ERR(__wt_tiered_switch(session, config));
     } else {
+        /*
+         * Here we expect all elements in tiers to be NULL. In certain scenarios tiered dhandle may
+         * come from the cache with initialized but invalid tiers. Make sure there are no tiers
+         * referenced and and free the memory.
+         */
+        for (i = 0; i < WT_TIERED_MAX_TIERS; ++i) {
+            if (tiered->tiers[i].tier != NULL)
+                (void)__wt_atomic_subi32(&tiered->tiers[i].tier->session_inuse, 1);
+            __wt_free(session, tiered->tiers[i].name);
+            tiered->tiers[i].tier = NULL;
+            tiered->tiers[i].name = NULL;
+        }
+
```

Comment:
If I understand the situation correctly, a previous call has failed leaving the dhandle in a bad state, and we're dealing with it here.  Is there a way instead to leave the dhandle in a sane state after the original failure?

FYI, the call to `__wt_free` clears its second argument, so setting `tiered->tiers[i].name = NULL` is redundant.

Reply:
Thank you for the feedback! Yes, I was thinking about it too. We should clean up the dhandle before leaving the failed call to prevent it from sitting in the cache with the invalid data. I'm working on finding a way to do it. It will probably be implemented as a part of the metadata unroll process.
This PR will not go ahead. I'm going to mark it as a draft for now. Thanks!